torchrl/objectives/iql.py

```diff
@@ -0,0 +1,255 @@
+                f"Losses shape mismatch: {log_prob.shape} and {min_q.shape}"
+            )
+        # state value
+        with torch.no_grad():
+            td_copy = tensordict.select(*self.value_network.in_keys).detach()
+            self.value_network(
+                td_copy,
+                params=self.value_network_params,
+            )
+            value = td_copy.get("state_value").squeeze(-1)  # assert has no gradient
+
+        exp_a = torch.exp((min_q - value) * self.temperature)
+        exp_a = torch.min(exp_a, torch.FloatTensor([100.0]).to(self.device))
+
+        # write log_prob in tensordict for alpha loss
+        tensordict.set("_log_prob", log_prob.detach())
+        return -(exp_a * log_prob).mean()
+
+    def _loss_value(self, tensordict: TensorDictBase) -> Tuple[Tensor, Tensor]:
+        # Min Q value
+        td_q = tensordict.select(*self.qvalue_network.in_keys)
+        td_q = vmap(self.qvalue_network, (None, 0))(
+            td_q, self.target_qvalue_network_params
+        )
+        min_q = td_q.get("state_action_value").min(0)[0].squeeze(-1)
+        # state value
+        td_copy = tensordict.select(*self.value_network.in_keys)
+        self.value_network(
+            td_copy,
+            params=self.value_network_params,
+        )
+        value = td_copy.get("state_value").squeeze(-1)
+        value_loss = loss_value_diff(min_q - value, self.expectile).mean()
+        return value_loss
+
+    def _loss_qvalue(self, tensordict: TensorDictBase) -> Tuple[Tensor, Tensor]:
+        obs_keys = self.actor_network.in_keys
+        tensordict = tensordict.select("reward", "done", "next", *obs_keys, "action")
+
+        with torch.no_grad():
+            next_td = step_mdp(tensordict).select(
+                *self.actor_network.in_keys
+            )  # next_observation ->
+            # observation
+            # select pseudo-action
+            # get state values
+            next_td = self.value_network(
+                next_td,
+                params=self.value_network_params,
+            )
+
+            state_value = next_td.get("state_value")
+
+        tensordict.set("next.state_value", state_value)
+        target_value = next_state_value(
+            tensordict,
+            gamma=self.gamma,
+            pred_next_val=state_value,
+        )
+        tensordict_expand = vmap(self.qvalue_network, (None, 0))(
+            tensordict.select(*self.qvalue_network.in_keys),
+            self.qvalue_network_params,
+        )
+        pred_val = tensordict_expand.get("state_action_value").squeeze(-1)
+        td_error = abs(pred_val - target_value)
+        loss_qval = (
+            distance_loss(
+                pred_val,
+                target_value.expand_as(pred_val),
+                loss_function=self.loss_function,
+            )
+            .sum(0)
+            .mean()
+        )
+        return loss_qval, td_error.detach().max(0)[0]
+
+
+def loss_value_diff(diff, expectile=0.8):
```

Comment:
is there a reason to make this public? Shouldn't we use a `staticmethod` decorator? Otherwise we can move it to some utils/functional file if the scope is broader.